pairinteraction/CMakeLists.txt

```diff
@@ -140,6 +140,31 @@ if (NOT MKL_FOUND AND PKG_CONFIG_FOUND)
   endif()
 endif()
 
+# https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-windows/2023-1/contents-of-the-redist-intel64-directory.html
+function(target_link_mkl TARGET)
+  if(MSVC AND TARGET MKL::MKL)
+    set(MKL_REDIST_DIR ${MKL_ROOT}/redist/intel64)
+    list(APPEND MKL_DLLS
+      ${MKL_REDIST_DIR}/mkl_avx.2.dll
+      ${MKL_REDIST_DIR}/mkl_avx2.2.dll
+      ${MKL_REDIST_DIR}/mkl_avx512.2.dll
+      ${MKL_REDIST_DIR}/mkl_core.2.dll
+      ${MKL_REDIST_DIR}/mkl_def.2.dll
+      ${MKL_REDIST_DIR}/mkl_${MKL_THREADING}.2.dll)
```

Comment:
I think it is better to use `${MKL_REDIST_DIR}/mkl_avx*.dll` etc. as it is done in the MKLConfig.cmake, instead of fixing the version of the DLLs to 2. To always copy the newest version of the DLLs, one could write for the foreach loop (untested):

```cmake
foreach(EXPRESSION ${MKL_DLLS})
  file(GLOB FOUND_FILES EXPRESSION)
  list(SORT FOUND_FILES)
  list(REVERSE FOUND_FILES)
  list(GET FILE 0 FOUND_FILES)
  add_custom_command(TARGET ${TARGET} POST_BUILD
     COMMAND ${CMAKE_COMMAND} -E copy_if_different ${FILE} $<TARGET_FILE_DIR:${TARGET}>)
endforeach()
```

Reply:
Should we add the DLLs "mkl_mc", "mkl_mc3", "libimalloc", and maybe the "mkl_vml_..." DLLs to be on the safe side? My fear is that depending on the processor on which pairinteraction runs, MKL wants to load some of these DLLs as well.

Reply:
1. We don't need to any of `mkl_mc.<version>.dll` or `mkl_mc3.<version>.dll` because these are only needed for processors that do not support AVX but only the older SSE instruction set. However, pairinteraction is now compiled with explicit AVX2 support, so any processor without AVX is unsupported anyway.

2. The `mkl_vml_*.<version>.dll` files are for the [Vector Mathematical Functions](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-1/vector-mathematical-functions.html) but we are not using any of those. This is a separate API.

3. We also do not at this moment try to [make Intel MKL use another memory allocator](https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-windows/2023-1/redefining-memory-functions.html) which is what `libimalloc.dll` is for.

    > Intel® oneAPI Math Kernel Library memory management by default uses standard C run-time memory functions to allocate or free memory. These functions can be replaced using memory renaming.

4. I would really like to avoid globbing because then we have to think about what to do if the glob expression returned something empty. Do we fail? But what if one of the DLLs we're trying to glob has been removed in that particular version, so the glob will always fail? The bottomline is that when we upgrade MKL we have to read the documentation anyway, so there is not really a point in complicating things. Also generally I prefer this failing early and loudly rather than late and silently.

Reply:
BTW, you proposal fails in the presence of versioned and unversioned DLLs because it will always pick the unversioned one this way.
```cmake
#!/usr/bin/env -S cmake -P
list(APPEND FOUND_FILES mkl_def.dll mkl_def.1.dll mkl_def.2.dll) # pretend that we globbed this
list(SORT FOUND_FILES)
list(REVERSE FOUND_FILES)
message(${FOUND_FILES})
list(GET FOUND_FILES 0 FILE)
message(${FILE})
```
prints
```console
$ ./test.cmake
mkl_def.dllmkl_def.2.dllmkl_def.1.dll
mkl_def.dll
```

Reply:
> The mkl_vml_*.<version>.dll files are for the [Vector Mathematical Functions](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-1/vector-mathematical-functions.html) but we are not using any of those. This is a separate API.

Isn't this used for some functions in eigen?

Reply:
Ah, yes, you're right, but only for array operations that we currently don't use. But we should indeed ship the DLLs in this case because in principle these functions could sneak in at some point.

| Code example              | MKL routines |
|---------------------------|--------------|
| `v2=v1.array().sin();`    | `v?Sin`      |
| `v2=v1.array().asin();`   | `v?Asin`     |
| `v2=v1.array().cos();`    | `v?Cos`      |
| `v2=v1.array().acos();`   | `v?Acos`     |
| `v2=v1.array().tan();`    | `v?Tan`      |
| `v2=v1.array().exp();`    | `v?Exp`      |
| `v2=v1.array().log();`    | `v?Ln`       |
| `v2=v1.array().sqrt();`   | `v?Sqrt`     |
| `v2=v1.array().square();` | `v?Sqr`      |
| `v2=v1.array().pow(1.5);` | `v?Powx`     |

https://eigen.tuxfamily.org/dox/TopicUsingIntelMKL.html

Reply:
Following this discussion I think just stupidly copying all of the DLLs is the easiest.
```patch
diff --git a/pairinteraction/CMakeLists.txt b/pairinteraction/CMakeLists.txt
index 1c7d777..1d4007d 100644
--- a/pairinteraction/CMakeLists.txt
+++ b/pairinteraction/CMakeLists.txt
@@ -144,13 +144,7 @@ endif()
 function(target_link_mkl TARGET)
   if(MSVC AND TARGET MKL::MKL)
     set(MKL_REDIST_DIR ${MKL_ROOT}/redist/intel64)
-    list(APPEND MKL_DLLS
-      ${MKL_REDIST_DIR}/mkl_avx.2.dll
-      ${MKL_REDIST_DIR}/mkl_avx2.2.dll
-      ${MKL_REDIST_DIR}/mkl_avx512.2.dll
-      ${MKL_REDIST_DIR}/mkl_core.2.dll
-      ${MKL_REDIST_DIR}/mkl_def.2.dll
-      ${MKL_REDIST_DIR}/mkl_${MKL_THREADING}.2.dll)
+    file(GLOB MKL_DLLS "${MKL_REDIST_DIR}/*.dll")
     if(MKL_THREADING STREQUAL "intel_thread")
       # Implementation detail of MKLConfig.cmake, the pkg-config files do not expose these variables
       list(APPEND MKL_DLLS ${OMP_DLL_DIR}/${OMP_DLLNAME})
```

Reply:
How much does this increase the file size? If the increase is small, I like this solution as it is perhaps the most robust thing that we can do. If MKL's "cluster libraries" are increasing the file size a lot, I would still prefer globing individual DLLs (https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-windows/2023-1/contents-of-the-redist-intel64-directory.html is using a placeholder for the version in the file names of the DLLs, so my proposed globing should be fine).

Reply:
So you mean we should glob *only* versioned DLLs and ignore unversioned?

Reply:
The cluster libraries are tiny, but the two `mkl_sycl*.dll` are quite large, see the listing in my previous comment: https://github.com/pairinteraction/pairinteraction/pull/149#discussion_r1190748794